src/cpp/core/Database.cpp

```diff
@@ -0,0 +1,635 @@
+                          options.host %
+                          options.port %
+                          options.database %
+                          options.user %
+                          options.password %
+                          safe_convert::numberToString(options.connectionTimeoutSeconds, "0"));
+
+         boost::shared_ptr<IConnection> pConnection(new Connection(soci::postgresql, connectionStr));
+         *pPtrConnection_ = pConnection;
+         return Success();
+      }
+      catch (soci::soci_error& error)
+      {
+         return DatabaseError(error);
+      }
+   }
+
+private:
+   boost::shared_ptr<IConnection>* pPtrConnection_;
+};
+
+Query::Query(const std::string& sqlStatement,
+             soci::session& session) :
+   statement_(session)
+{
+   // it's possible that prepare can throw a database exception, but we
+   // do not want to surface errors until execute() is called
+   try
+   {
+      statement_.alloc();
+      statement_.prepare(sqlStatement);
+   }
+   catch (soci::soci_error& error)
+   {
+      prepareError_ = error;
+   }
+}
+
+RowsetIterator Rowset::begin()
+{
+   if (query_)
+      return RowsetIterator(query_.get().statement_, row_);
+
+   return end();
+}
+
+RowsetIterator Rowset::end()
+{
+   return RowsetIterator();
+}
+
+Connection::Connection(const soci::backend_factory& factory,
+                       const std::string& connectionStr) :
+   session_(factory, connectionStr)
+{
+}
+
+Query Connection::query(const std::string& sqlStatement)
+{
+   return Query(sqlStatement, session_);
+}
+
+Error Connection::execute(Query& query,
+                          bool* pDataReturned)
+{
+   if (query.prepareError_)
+      return DatabaseError(query.prepareError_.get());
+
+   try
+   {
+      query.statement_.define_and_bind();
+      bool result = query.statement_.execute(true);
+
+      if (pDataReturned)
+         *pDataReturned = result;
+
+      query.statement_.bind_clean_up();
+
+      return Success();
+   }
+   catch (soci::soci_error& error)
+   {
+      return DatabaseError(error);
+   }
+}
+
+Error Connection::execute(Query& query,
+                          Rowset& rowset)
+{
+   if (query.prepareError_)
+      return DatabaseError(query.prepareError_.get());
+
+   try
+   {
+      query.statement_.define_and_bind();
+      query.statement_.exchange_for_rowset(soci::into(rowset.row_));
+      query.statement_.execute(false);
+
+      rowset.query_ = query;
+
+      return Success();
+   }
+   catch (soci::soci_error& error)
+   {
+      return DatabaseError(error);
+   }
+}
+
+Error Connection::executeStr(const std::string& queryStr)
+{
+   try
+   {
+      // SOCI backends do not necessarily support running multiple statements
+      // in one invocation - to work around this, we split any passed in SQL
+      // into one invocation per SQL statement (delimited by ;)
+      std::vector<std::string> queries;
+      boost::split(queries, queryStr, boost::is_any_of(";"));
```

Comment:
Will we need to worry about queries which include `;` as part of e.g. strings or something similar?

Reply:
Yes, this is a good point. My next commit will fix this by using `regex_split`